Interface physics across domains in DG

lindsayad opened the pull request:
Added preliminary ability to interface different physics domains using DGKernels and boundary restriction. Attached are two crude pictures showing the interfacing of two domains with different diffusion coefficients, assuming continuity of _u and continuity of flux. One case has _d_coeff_left = 1 and _d_coeff_right = 2. The other case has _d_coeff_left = 1 and _d_coeff_right = 3.

I was having problems rebasing on top of the development branch, but I was able to merge my changes into the development branch. If that's a problem, please let me know. I'm not yet a git expert! I hope that this will be the start of something useful for other people. I'm really excited about using it for my own needs, and I'm perhaps even more excited to try and make a contribution to the core Moose framework.

![dg_interface_d1_equals_1_d2_equals_2](https://cloud.githubusercontent.com/assets/10248304/11051433/4b855c2e-871d-11e5-9e7e-bc9e172a7aa4.png)
![dg_interface_d1_equals_1_d2_equals_3](https://cloud.githubusercontent.com/assets/10248304/11051444/606f6b84-871d-11e5-8a27-b41c957ab811.png)

Closes #869

lindsayad commented:
It looks like I also included some undesirable file additions/changes in this PR that don't have anything to do with my DGInteface implementation. I'm heading home now, but I can fix this tomorrow.

andrsd commented:
Go look at https://github.com/andrsd/moose/tree/lindsayad/869 if that's your work. I rebased you branch on top of current devel.

Rules of thumb for MOOSE development:
 1. Always branch from devel
 2. Do not merge devel/master into your feature branch

dschwen commented:
@lindsayad I tried to give your PR a more descriptiove title. Please check if it fits. Also: the issue number goes in the description.

lindsayad commented:
David, that is my work. 

Yea, when I was looking at the contributing page last night, I realized I'd made a mistake in branching from master. I should have read that page before trying to introduce modifications!

What's the problem with making periodic merges from devel (or master) into my feature branch during development? For instance if I want to keep up with the newest moose features during my development process? Just messes up the history or what?

Returning to this PR, how should I proceed? I realized that I pushed some file additions/modification changes that aren't relevant to this PR request (or desirable to have in the moose framework in general). Should I fetch and merge andrsd/moose lindsayad/869 into my local feature branch and then remove the changes I want to remove? Or do I need to do some rebasing again?

Sorry guys. I promise any future development work will be less sloppy and follow the Moose devel rules!

lindsayad commented:
@dschwen I like your title better!

jwpeterson commented:
> What's the problem with making periodic merges from devel (or master) into my feature branch during development?

There's no technical problem with this, i.e. git handles it just fine, but there's a human problem.  If I look at your PR and it contains dozens of commits that were merged in from devel (or master, as in the case above) it's much harder to review than it needs to be.

andrsd commented:
> Returning to this PR, how should I proceed?

Grab my branch via git fetch and force push it into this PR, then we can start reviewing of your code.

lindsayad commented:
@jwpeterson Yes, I see now that instead submitting with 8 commits, I was submitting with well over 250. I can see that that is a big pain!

@andrsd Hopefully, this is what you were looking for.

andrsd commented:
@lindsayad You will need to format the code so it complies which the coding standards. Things I saw to fix:

1. Pad operators with spaces, e.g.  `1/3` would be `1 / 3'.
2. Correct indentation is to have curly braces on individual lines, e.g.:
```
if (cond)
{
  ...
}
```
3. You have some commented out code in your PR which can be removed

lindsayad commented:
Something I should note is that currently the interfacing will only work if the "correct" variable is chosen to be the master variable for the interface kernel. I am pretty certain that this stems from this logic which I took from ComputeResidual(Jacobian)Thread::onInternalSide()

  if ((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) || (neighbor->level() < elem->level()))
......
.....l
computeResidual()

I haven't yet thought of an intelligent way to make sure this works regardless of the selection of the master variable.

lindsayad commented:
Looking at the checks, it looks like I'm getting a lot of errors like the following when the framework is being built:

DGInterface.C:61:72: error: invalid use of incomplete type 'class Assembly'

1) What does this mean? My hypothesis is that this would go away if I included Assembly.h in DGInterface.h and DGInterface.C, but that's just based on intuition, not on sound C++ knowledge.

2) How come I don't see those errors when I'm compiling the framework locally?

jwpeterson commented:
> Looking at the checks, it looks like I'm getting a lot of errors like the following when the framework is being built:
> 
> DGInterface.C:61:72: error: invalid use of incomplete type 'class Assembly'
> 
> 1) What does this mean? My hypothesis is that this would go away if I included Assembly.h in DGInterface.h and DGInterface.C, but that's just based on intuition, not on sound C++ knowledge.

You need to #include "Assembly.h" in your class.

> 2) How come I don't see those errors when I'm compiling the framework locally?

Your MOOSE is out of date.  We've moved around header files since the last time you did an update.

jwpeterson commented on framework/include/dgkernels/DGInterface.h:
```diff
@@ -0,0 +1,56 @@
+template<>
+InputParameters validParams<DGInterface>();
+
+/* Objects whose type inherits from DGInterface must currently define the master variable to be
```
Class comments should use Doxygen format:
```
/**
 * Comments comments comments.
 */
```

jwpeterson commented on framework/include/dgkernels/DGInterface.h:
```diff
@@ -0,0 +1,56 @@
+public:
+  DGInterface(const InputParameters & parameters);
+
+  std::vector<dof_id_type> _connected_dof_indices;
```
Do you intend for `_connected_dof_indices`, `_Kne`, and `_Kee` to be public?  They are named in the way we usually name protected variables.

lindsayad replied:
I didn't end up actually using those variables, so I've deleted them.

jwpeterson commented on framework/src/base/ComputeJacobianThread.C:
```diff
@@ -222,6 +238,46 @@ ComputeJacobianThread::onInternalSide(const Elem *elem, unsigned int side)
+    elem_id = elem->id(),
+    neighbor_id = neighbor->id();
+
+  if ((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) || (neighbor->level() < elem->level()))
```
So... what happens if there are different levels of refinement in the elements on either side of the interface?  Is this all designed to work with adaptivity?

lindsayad replied:
It hasn't been a focus of mine, but I suppose that I would like the code to be as general as possible, thus I would like for it to work with adaptivity. I basically took those checks verbatim from the onInternalSide method. By cursory inspection they look like they're taking care of adaptivity, but I could definitely be wrong.

jwpeterson commented on framework/src/dgkernels/DGDiffusionInterface.C:
```diff
@@ -0,0 +1,90 @@
+{
+  Real r = 0;
+
+  switch (type)
```
The difference in the two cases seems to be a minus sign and which test function you multiply by.  So I'd initialize `r` with:
```
r = 0.5 * (-_D * _grad_u[_qp] * _normals[_qp] + -_D_neighbor * _grad_neighbor_value[_qp] * _normals[_qp]);
```

then return either `r * _test[_i][_qp]` or `-r * _test_neighbor[_i][_qp]` directly from the cases.

jwpeterson commented on framework/src/dgkernels/DGKernel.C:
```diff
@@ -29,15 +29,12 @@
 #include "libmesh/libmesh_common.h"
 #include "libmesh/quadrature.h"
 
-const BoundaryID DGKernel::InternalBndId = 12345;
```
Hmm... did you mean to delete this?  Seems like it might be important.

jwpeterson replied:
Also it's the same combination as my luggage!

lindsayad replied:
As far as I could tell this member served no purpose. I took it to mean that a developer in the past had premonitions about adding BoundaryRestriction but never got around to it. Before this PR DGKernels were incapable of boundary restriction, so what use would this seemingly miscellaneous BoundaryID have? That was my take. I very well could have missed its importance however.

jwpeterson commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,145 @@
+// libMesh includes
+#include "libmesh/quadrature.h"
+
+#include <cmath>
```
I don't think you need `<cmath>` but maybe I missed something.

lindsayad replied:
I don't think I need it either.

YaqiWang commented:
I like this change overall! But we need to put in tests for this.

lindsayad commented:
I'm in a time crunch for getting my research done (trying to graduate this spring or summer). I need block restricted DG's for my own application, so I'm starting to code some things up in the framework, with plans of modelling the subdomain restriction in the onInternalSide() methods of ComputeResidualThread and ComputeJacobianThread after the onInternalSide method in ComputeUserObjectThread. I think this is something of interest to the framework, as Yaqi has indicated. I could include the changes I'm looking to make in this PR if people think that's a good idea, or if people are tired of reviewing my code, I understand :-) (Although I think I'm improving with time!)

permcody commented:
Alex, We'd be happy to review your code and offer advice. Looking forward
to seeing what you come up with.

On Tue, Dec 8, 2015 at 6:55 PM Alex Lindsay <notifications@github.com>
wrote:

> I'm in a time crunch for getting my research done (trying to graduate this
> spring or summer). I need block restricted DG's for my own application, so
> I'm starting to code some things up in the framework, with plans of
> modelling the subdomain restriction in the onInternalSide() methods of
> ComputeResidualThread and ComputeJacobianThread after the onInternalSide
> method in ComputeUserObjectThread. I think this is something of interest to
> the framework, as Yaqi has indicated. I could include the changes I'm
> looking to make in this PR if people think that's a good idea, or if people
> are tired of reviewing my code, I understand :-) (Although I think I'm
> improving with time!)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-163081480>.
>

permcody commented:
@lindsayad - are you planning on coming back to this? It looks like it's pretty close to being merged but there are still a few developer comments that should be resolved.

lindsayad commented:
Yes, I'm working on the DGKernel block restriction right now. I think I should have something by the end of the day

lindsayad commented:
![verification_of_block_restriction](https://cloud.githubusercontent.com/assets/10248304/11732100/0463b452-9f6f-11e5-984f-2cb1700b8f04.png)
[sub_2_two_var_reg_dg_block_restric_phys_test.txt](https://github.com/idaholab/moose/files/58891/sub_2_two_var_reg_dg_block_restric_phys_test.txt)

Attached picture is of last time step generated with attached input file. Shows block restriction of dgs.

lindsayad commented:
Alright, added block restriction, so there's some new code to review. I hope that I addressed all the developer comments from the previous (dg_interface) commit.

permcody commented:
@lindsayad - It looks like you forgot to check in your input file "dg_block_restric.i". Can you see the test results? For some reason they didn't get reported back on this page:
https://www.moosebuild.org/job/5345/

lindsayad commented:
When I first pushed the last commit, tests was looking for 
"dg_block_restric.i" which it couldn't find because the file is named 
"dg_block_restrict.i". When I fixed this and pushed again, for some 
reason the moose tests didn't appear (e.g. build modules, build 
framework, pre-check, etc.) and they still haven't appeared. I don't 
know what's going on.

On 12/17/2015 04:19 PM, Cody Permann wrote:
>
> @lindsayad <https://github.com/lindsayad> - It looks like you forgot 
> to check in your input file "dg_block_restric.i". Can you see the test 
> results? For some reason they didn't get reported back on this page:
> https://www.moosebuild.org/job/5345/
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/idaholab/moose/pull/5943#issuecomment-165584400>.
>

permcody commented:
I'll have @brianmoose look into it. I'm not sure why this is happening. You
can see the results directly on moosebuild.org
On Thu, Dec 17, 2015 at 4:20 PM Alex Lindsay <notifications@github.com>
wrote:

> When I first pushed the last commit, tests was looking for
> "dg_block_restric.i" which it couldn't find because the file is named
> "dg_block_restrict.i". When I fixed this and pushed again, for some
> reason the moose tests didn't appear (e.g. build modules, build
> framework, pre-check, etc.) and they still haven't appeared. I don't
> know what's going on.
>
> On 12/17/2015 04:19 PM, Cody Permann wrote:
> >
> > @lindsayad <https://github.com/lindsayad> - It looks like you forgot
> > to check in your input file "dg_block_restric.i". Can you see the test
> > results? For some reason they didn't get reported back on this page:
> > https://www.moosebuild.org/job/5345/
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-165584400>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-165612943>.
>

brianmoose commented:
Yeah, github had an error delivering the payload for this for some reason.
It is running now.

lindsayad commented:
So it looks like my tests pass now, but now this is failing:

utils/random.threads_verification_uo.......................................................... FAILED (CRASH)

I don't know what this might be about

permcody commented:
Just ignore that. That's our random failing test. For real...
On Thu, Dec 17, 2015 at 5:02 PM Alex Lindsay <notifications@github.com>
wrote:

> So it looks like my tests pass now, but now this is failing:
>
> utils/random.threads_verification_uo..........................................................
> FAILED (CRASH)
>
> I don't know what this might be about
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-165619987>.
>

permcody commented:
@lindsayad - Thanks for working on this!
@idaholab/moose-developers - please review

permcody commented:
@lindsayad - Thanks for working on this!
@idaholab/moose-developers - please review

permcody commented:
@andrsd - This PR has been sitting here for awhile. Do you believe these changes are reasonable to incorporate into the framework baring any cosmetic formatting issues and the current merge conflict? @friedmud, care to weigh in?

andrsd commented:
> Do you believe these changes are reasonable to incorporate into the framework

This was supposed to implement DGBCs (IIRC), so if we want that, somebody needs to polish the code up ;-)

lindsayad commented:
David,

Can you explicitly say what's missing? I feel like this PR addresses the issue outlined in #869, which asks for capability to restrict a dgkernel to an internal side set. What are you looking for?

andrsd commented on framework/include/dgkernels/DGKernelWarehouse.h:
```diff
@@ -38,14 +39,39 @@ class DGKernelWarehouse : public Warehouse<DGKernel>
+  bool hasActiveDGKernels(unsigned int var) const
+    {
+      return _active_var_dg_kernels.find(var) != _active_var_dg_kernels.end();
+    }
```
Incorrect indentation

lindsayad replied:
Check

andrsd commented on framework/include/dgkernels/DGKernelWarehouse.h:
```diff
@@ -38,14 +39,39 @@ class DGKernelWarehouse : public Warehouse<DGKernel>
+    {
+      mooseAssert(_active_var_dg_kernels.find(var) != _active_var_dg_kernels.end(), "No active dg_kernels");
+      return _active_var_dg_kernels.find(var)->second;
+    }
```
Incorrect indentation

lindsayad replied:
Check

andrsd commented on framework/src/base/ComputeJacobianThread.C:
```diff
@@ -222,6 +239,47 @@ ComputeJacobianThread::onInternalSide(const Elem *elem, unsigned int side)
+    _fe_problem.reinitMaterialsNeighbor(neighbor->subdomain_id(), _tid);
+
+    // Set the active boundary id so that BoundaryRestrictable::_boundary_id is correct
+    _fe_problem.setCurrentBoundaryID(bnd_id);
```
Can you refresh my memory why we need to set the current boundary ID?

lindsayad replied:
After some investigation, I don't really see why I need to do this. I modeled this after https://github.com/idaholab/moose/blob/devel/framework/src/base/ComputeJacobianThread.C#L171 but I can't see it's importance here (it's not immediately obvious to me that it's necessary in the linked line either but I just may not have dug deep enough)

andrsd replied:
I checked the code and `BoundaryRestrictable` relies on this value to be set. So this is correct, since `DGKernel` is now `BoundaryRestrictable`.

andrsd commented on framework/src/base/ComputeResidualThread.C:
```diff
@@ -91,6 +91,24 @@ ComputeResidualThread::subdomainChanged()
+      ++id_it)
+  {
+    std::vector<DGKernel *> interface_dgks;
+    _sys.getDGKernelWarehouse(_tid).activeInterfaceDG(*id_it, interface_dgks);
```
Typically the warehouse interface returns a reference to a container of objects, it is not filling up references.
Is that still correct for our warehouse refactoring @aeslaughter ?

lindsayad replied:
@andrsd 
I modeled this after an earlier version of the onBoundary method, which had this:

  std::vector<IntegratedBC *> bcs;
  _sys.getBCWarehouse(_tid).activeIntegrated(bnd_id, bcs);

It looks like the onBoundary method has been updated to look like the other methods (e.g. returning a reference to a container like you say), so I will fix my method to be consistent with the rest.

andrsd replied:
Got it.
That might the reason for the merge conflict.

andrsd commented on framework/src/base/ComputeResidualThread.C:
```diff
@@ -166,26 +187,65 @@ ComputeResidualThread::onInternalSide(const Elem *elem, unsigned int side)
+    for (std::vector<DGKernel *>::const_iterator it = dgs.begin(); it != dgs.end(); ++it)
+    {
+      DGKernel * dg = *it;
+      if ( dg->hasBlocks(neighbor->subdomain_id()) )
```
No padding around the boolean statement

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+InputParameters validParams<DGInterface>()
+{
+  InputParameters params = validParams<DGKernel>();
+  params.addRequiredCoupledVar("neighbor_var","The variable on the other side of the interface.");
```
Space after a `,`

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+
+DGInterface::DGInterface(const InputParameters & parameters) :
+    DGKernel(parameters),
+    _neighbor_var(*getVar("neighbor_var",0)),
```
Space after a `,`

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+                                       _assembly.residualBlockNeighbor(_neighbor_var.number());
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+      for (_i = 0; _i < test_space.size(); _i++)
```
Indentation

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+      for (_i = 0; _i < test_space.size(); _i++)
+        for (_j = 0; _j < loc_phi.size(); _j++)
+          Kxx(_i,_j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);
```
Space after a `,`

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+
+
+void
+DGInterface::computeOffDiagElemNeighJacobian(Moose::DGJacobianType type,unsigned int jvar)
```
Space after a `,`

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+    for (_qp=0; _qp<_qrule->n_points(); _qp++)
+      for (_i=0; _i<test_space.size(); _i++)
+        for (_j=0; _j<loc_phi.size(); _j++)
+          Kxx(_i,_j) += _JxW[_qp]*_coord[_qp]*computeQpOffDiagJacobian(type, jvar);
```
Spaces around `=` and  `<`.
Space after a `,`

lindsayad replied:
Check

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+    // Compute neighbor-neighbor Jacobian
+    computeOffDiagElemNeighJacobian(Moose::NeighborNeighbor,jvar);
+  }
+}
```
Space after a ',' - fix in the whole body.

lindsayad replied:
Check

andrsd commented on test/src/bcs/InflowBC.C:
```diff
@@ -0,0 +1,29 @@
+    InputParameters params = validParams<IntegratedBC>();
+    params.addRequiredParam<RealVectorValue>("velocity", "The velocity vector");
+    params.addRequiredParam<Real>("inlet_conc", "The inlet concentration");
+    return params;
```
Indentation

lindsayad replied:
Check

andrsd commented on test/src/bcs/InflowBC.C:
```diff
@@ -0,0 +1,29 @@
+  IntegratedBC(parameters),
+
+  _velocity(getParam<RealVectorValue>("velocity")),
+  _inlet_conc(getParam<Real>("inlet_conc"))
```
Indent by 4 spaces

andrsd commented on test/src/bcs/OutflowBC.C:
```diff
@@ -0,0 +1,28 @@
+{
+    InputParameters params = validParams<IntegratedBC>();
+    params.addRequiredParam<RealVectorValue>("velocity", "The velocity vector");
+    return params;
```
Indent by 2 spaces

andrsd commented on test/src/bcs/OutflowBC.C:
```diff
@@ -0,0 +1,28 @@
+OutflowBC::OutflowBC(const InputParameters & parameters) :
+  IntegratedBC(parameters),
+
+  _velocity(getParam<RealVectorValue>("velocity"))
```
Indent by 4 spaces

andrsd commented on test/src/bcs/RobinBC.C:
```diff
@@ -0,0 +1,40 @@
+RobinBC::computeQpResidual()
+{
+  return _test[_i][_qp] * 2. * _u[_qp];
+  // return _test[_i][_qp] * -2. * _grad_u[_qp] * _normals[_qp];
```
Delete

andrsd commented on test/src/bcs/RobinBC.C:
```diff
@@ -0,0 +1,40 @@
+RobinBC::computeQpJacobian()
+{
+  return _test[_i][_qp] * 2. * _phi[_j][_qp];
+  // return _test[_i][_qp] * -2. * _grad_phi[_j][_qp] * _normals[_qp];
```
Delete

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+  switch (type)
+  {
+    case Moose::Element:
+      if ( (_velocity * _normals[_qp]) >= 0.0)
```
No space after `(`

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+      break;
+
+    case Moose::Neighbor:
+      if ( (_velocity * _normals[_qp]) >= 0.0)
```
No space after `(`

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+
+  switch (type)
+  {
+
```
No need for an empty line

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+  {
+
+    case Moose::ElementElement:
+      if ( (_velocity * _normals[_qp]) >= 0.0)
```
No space after `(`

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+      break;
+
+    case Moose::NeighborElement:
+      if ( (_velocity * _normals[_qp]) >= 0.0)
```
No space after `(`

andrsd commented on test/src/dgkernels/DGAdvection.C:
```diff
@@ -0,0 +1,77 @@
+      break;
+
+    case Moose::NeighborNeighbor:
+      if ( (_velocity * _normals[_qp]) >= 0.0)
```
No space after `(`

andrsd commented on test/src/dgkernels/DGDiffusionInterface.C:
```diff
@@ -0,0 +1,90 @@
+{
+  InputParameters params = validParams<DGInterface>();
+  params.addParam<Real>("D",1.,"The diffusion coefficient.");
+  params.addParam<Real>("D_neighbor",1.,"The neighboring diffusion coefficient.");
```
Space after a `,`

andrsd commented on test/src/dgkernels/DGDiffusionInterface.C:
```diff
@@ -0,0 +1,90 @@
+
+  switch (type)
+  {
+
```
Remove

andrsd commented on test/src/dgkernels/DGDiffusionInterface.C:
```diff
@@ -0,0 +1,90 @@
+Real
+DGDiffusionInterface::computeQpOffDiagJacobian(Moose::DGJacobianType type, unsigned int jvar)
+{
+
```
Remove

andrsd commented on framework/src/base/ComputeFullJacobianThread.C:
```diff
@@ -159,12 +159,48 @@ ComputeFullJacobianThread::computeFaceJacobian(BoundaryID bnd_id)
+      for (std::vector<DGKernel *>::const_iterator dg_it = dgs.begin(); dg_it != dgs.end(); ++dg_it)
+      {
+        DGKernel * dg = *dg_it;
+        if ( dg->isImplicit() && dg->hasBlocks(neighbor->subdomain_id()) )
```
No spaces around the boolean

andrsd commented on framework/src/base/ComputeJacobianThread.C:
```diff
@@ -75,9 +75,27 @@ ComputeJacobianThread::computeFaceJacobian(BoundaryID bnd_id)
+  for (std::vector<DGKernel *>::const_iterator it = dgs.begin(); it != dgs.end(); ++it)
+  {
+    DGKernel * dg = *it;
+    if ( dg->isImplicit() && dg->hasBlocks(neighbor->subdomain_id()) )
```
No space around the boolean

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+{
+  const VariableTestValue & test_space = ( type == Moose::ElementElement || type == Moose::ElementNeighbor ) ?
+                                         _test : _test_neighbor;
+  const VariableTestValue & loc_phi = ( type == Moose::ElementElement || type == Moose::NeighborElement ) ?
```
No spaces inside `(...)`

andrsd commented on framework/src/dgkernels/DGInterface.C:
```diff
@@ -0,0 +1,143 @@
+void
+DGInterface::computeJacobian(unsigned int jvar)
+{
+
```
Remove

andrsd commented:
> What are you looking for?

Functionality is correct, just fix the code formatting.

lindsayad commented:
Ah, so this PR has sat here long enough that the warehouse system has changed. Now I have to make my brain hurt again :-(

permcody commented:
I blame @aeslaughter and @andrsd. If you run into problems with any of the
new stuff, please let us know and we'll lend a hand. I'll get this merged
as soon as it's ready.

On Thu, Jan 7, 2016 at 4:42 PM Alex Lindsay <notifications@github.com>
wrote:

> Ah, so this PR has sat here long enough that the warehouse system has
> changed. Now I have to make my brain hurt again :-(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-169842119>.
>

Commit:
Implemented coupling of physics across domains. (#869)

Summary of work done for #869: The majority of the heavy lifting is done in the
InterfaceKernel class. InterfaceKernel is essentially a re-implentation of DGKernel, but
it's designed for coupling different variables across an interface. Thus
it has some important differences with DGKernel. Here are a few:

 - computeJacobian is called only for Moose::ElementElement and
   Moose::NeighborNeighbor. This because residual and residualNeighbor
   blocks are actually for different variables, meaning ElementNeighbor
   and NeighborElement should actually be in off-diagonal blocks of the
   Jacobian
 - The ElementNeighbor and NeighborElement Jacobian coupling of the two residual
   variables are accomplished in the computeOffDiagJacobian
   method. Coupling of additional variables is obviously also
   accomplished in the OffDiagJacobian block as well
 - In order to get InterfaceKernel to work, the onInterface method had to be
   added to the ComputeResidualThread and ComputeJacobianThread
   classes. The computeInternalInterfaceJacobian method was added to
   ComputeFullJacobianThread
 - InterfaceDiffusion in test/ was created to demonstrate and test the
   InterfaceKernel capabilities
 - In FEProblem::addMaterial, set _neighbor parameter equal to true for
   neighbor materials such that the coupled variables receive the correct
   values
 - InterfaceKernel is now it's own system with its own directory and input
   file syntax

Also Corrects face and neighbor shape functions for DG. (#6041)

Implements block restriction of DGKernels. (#6042)

Summary of work done for #6042:

 - Similar to ComputeUserObjectThread, in order to make sure that we
   don't try and compute block restricted dg_kernels at an interface
   between subdomains that don't both contain the dg_kernel object, the
   check dg->hasBlocks is added to onInternalSide in
   ComputeResidualThread and ComputeJacobianThread. The neighbor element
   is also passed to the computeInternalFaceJacobian methods in
   ComputeJacobianThread and ComputeFullJacobianThread
 - Also created tests for interface_kernels and block restriction of dg_kernels

To wrap: this PR addresses issues #869, #6041, #6042

lindsayad commented:
How often does git merge make a mistake? See attached ComputeResidualThread.C (cast in txt format so I can attach). If I keep HEAD (which represents moose/devel) and remove my changes, I reproduce all the code for the onInternalSide method but it now has my new method name (onInterface)!!! Does git only check out to a certain column number to determine if a line of code has changed? If that is the case, then I could see how this happened since the method names are very similar:

ComputeResidualThread::onInternalSide
ComputeResidualthread::onInterface

[ComputeResidualThread.txt](https://github.com/idaholab/moose/files/83283/ComputeResidualThread.txt)

permcody commented:
The answer is never. If changes occur on the same line git won't merge
those file automatically. Instead you get a conflict and git puts the lines
from both versions in the conflicting file with annotations and makes you
figure it out.
On Fri, Jan 8, 2016 at 2:18 PM Alex Lindsay <notifications@github.com>
wrote:

> How often does git merge make a mistake? See attached
> ComputeResidualThread.C (cast in txt format so I can attach). If I keep
> HEAD (which represents moose/devel) and remove my changes, I reproduce all
> the code for the onInternalSide method but it now has my new method name
> (onInterface)!!! Does git only check out to a certain column number to
> determine if a line of code has changed? If that is the case, then I could
> see how this happened since the method names are very similar:
>
> ComputeResidualThread::onInternalSide
> ComputeResidualthread::onInterface
>
> ComputeResidualThread.txt
> <https://github.com/idaholab/moose/files/83283/ComputeResidualThread.txt>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-170126189>.
>

lindsayad commented:
Alright, hopefully I don't need to make any more functional changes to this code because my brain hurts a lot and I just may cry haha (but seriously I might cry).

@aeslaughter @andrsd I have to say that the warehouse changes are very elegant. If they had been around the first time I coded in the dg boundary and block restriction, it would have really reduced the amount of work!

lindsayad commented:
@permcody As far as I'm concerned this PR is ready again, but I'm sure the developers may want to look at the changes

andrsd commented on framework/src/dgkernels/DGKernelWarehouse.C:
```diff
@@ -0,0 +1,123 @@
+  _active_var_dg_kernels.clear();
+
+    // FIXME: add startTime/stopTime to DGKernel
+//    if (dg_kernel->startTime() <= t + (1e-6 * dt) && dg_kernel->stopTime() >= t + (1e-6 * dt))
```
I assume, we do not want this line here...

friedmud replied:
I still want to check this.  Cloning it now...

On Mon, Jan 11, 2016 at 12:18 PM David Andrs <notifications@github.com>
wrote:

> In framework/src/dgkernels/DGKernelWarehouse.C
> <https://github.com/idaholab/moose/pull/5943#discussion_r49348674>:
>
> > +
> > +void
> > +DGKernelWarehouse::addInterfaceDG(const std::set<BoundaryID> & boundary_ids, MooseSharedPointer<DGKernel> & interface_dg)
> > +{
> > +  for (std::set<BoundaryID>::const_iterator it = boundary_ids.begin(); it != boundary_ids.end(); ++it)
> > +    _interface_dg_kernels[*it].push_back(interface_dg.get());
> > +}
> > +
> > +void
> > +DGKernelWarehouse::updateActiveDGKernels(Real /*t*/, Real /*dt*/, unsigned int subdomain_id)
> > +{
> > +  _active_internal_dg_kernels.clear();
> > +  _active_var_dg_kernels.clear();
> > +
> > +    // FIXME: add startTime/stopTime to DGKernel
> > +//    if (dg_kernel->startTime() <= t + (1e-6 * dt) && dg_kernel->stopTime() >= t + (1e-6 * dt))
>
> I assume, we do not want this line here...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943/files#r49348674>.
>

lindsayad replied:
The dgkernel warehouse class shouldn't exist anymore. My fault

friedmud commented on framework/src/base/Assembly.C:
```diff
@@ -1547,3 +1547,9 @@ Assembly::clearCachedJacobianContributions()
 }
+
+bool
+Assembly::isBlockDiagonal()
```
I don't understand why this was added - it doesn't seem to be getting called anywhere in your PR....

lindsayad replied:
Hmm, I think this might have been necessary before the warehouse system was changed. But yea now I'm not using it, so it can go!

friedmud commented:
I'm writing up a big thing on why this implementation won't exactly work the way it is.

Look for that a little later tonight ;-)

friedmud commented:
DG Interface

Ok.  You've done a lot of good work here... but there is a big issue.

The problem is with your whole plan with "master" variable's always being on the side with the lower element ID.  That doesn't work.  In general, the element IDs will be completely jumbled... there won't be a variable on one side or the other that is always on the lower element ID.  That whole idea needs to be abandoned ;-)

Actually, it's even worse than that... because of Mesh Adaptivity.  Our first swipe at this might actually need to disallow Mesh Adaptivity (that should probably be done with DG anyway... because stateful material properties don't work properly with DG and Mesh Adaptivity anyway!).  You'll see why in a second.

The problem that you’re running into is this line:

```(neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) || (neighbor->level() < elem->level())```

I’m not sure if you’ve deciphered what’s up with that line so here are the highlights:

1.  It makes it so that you only ever integrate over one side one time via ```elem_id < neighbor_id```.  Without that you would hit each side twice (once when you go through it on each element connect to the side).
2. Ensures that you integrate over the “smaller” elements on either side of a “hanging node” interface with adapted mesh via ```neighbor->level() < elem->level()```.  A higher “level” means _smaller_ elements… so if we’re on the side with smaller elements we need to integrate over this side _now_ (regardless of element ID).

Now… why does this not work with what you’re doing?  Because internal interfaces like this need to have directionality (i.e. they need to have a “sense” - a predictable normal vector).  You tried to impose directionality by having a “master” variable that’s on the lower element ID… but as I mentioned before, that doesn’t work.  So how do you impose directionality on an internal interface?

In Cubit you can choose a block ID that an internal boundary condition is “With Respect To”.  When you do that, it places the sideset _only_ on the sides of the elements of that block ID.  This makes it so that there is “directionality” to the sideset… i.e…. it gives it a consistent outward facing normal.  It accomplishes what you were trying to do with the “master” variable… because now there will always be one block that is the “element” and the other one will always be the “neighbor”…. guaranteed.

(from here on I’m using the “master” side to mean the block the sideset is “with respect to”)

Here’s the catch.  When you do that, it only places the boundary ID on the outside of the elements of the master side.  Meaning… if you were to query that same side from the “neighbor” element IT WOULD NOT HAVE A BOUNDARY ID ON IT AT ALL!  That means that we can no longer arbitrarily choose to integrate from either side of an internal interface… we _always_ have to integrate from the “master” element side.

That seems ok - and will actually go along with most of the rest of your code just fine… but it breaks down in mesh adaptivity.  If the neighbor element is refined but the master element is not… then we actually need to integrate from the “neighbor” side…  but there is no way to know that now because the neighbor side doesn’t even have the boundary ID on it.

Anyway… this is all somewhat too complicated to put in text like this.  Hopefully you’ve followed what I’ve said so far.

Here are my recommendations:

1.  Rename DG Interfaces to just `InterfaceKernels` (this is unrelated, but there is no reason for “DG” to be part of the name).
2. Require that internal boundaries be specified “With respect to” a block (the “master” block).  That way there is only _one_ element on the interface that will have that boundary ID... and it will be the "master" element.
3. The variable that lives on the “master” block will be the “master” variable
4. Always integrate from the “master” side

One really big upside is that this will give you a consistent normal vector (it will always point "out" from the master domain) which is important for a lot of physics interfaces.

Let me know if any of that doesn't make sense!

lindsayad commented:
That was all very clear Derek. Thanks. It could be a few days before I get to it, but I will work on implementing the changes you recommended (which all make good sense!).

friedmud commented:
@lindsayad awesome!

Now that I've thought a bit more about it... I'm not sure there is much you have to change.  Since the boundary ID will only exist on the master side... your code might work just as it is if you remove that check...'
Like I say: it won't work with mesh adaptivity, but that's something to think about later.

lindsayad commented:
@friedmud After removing the check (well I left the check of neighbor->active(); without doing any digging that still seems like a reasonable check (although I don't know whether it's necessary)), I can now put my interface_kernel on either subdomain and get the right result using my coupled_value_coupled_flux.i test.

lindsayad commented:
This is probably in the developer slides, but do you guys prefer PR's to be a single commit? I'm leery of continuing to squash into one commit locally and running push -f in the chance that someone else is working on the PR code (do people do that? If not, then I guess it's fine to keep squashing into one commit).

friedmud commented:
@lindsayad Great!  I'll take a look at the updated code soon (probably tomorrow).

A couple of things on squashing:

1.  There is no problem with force pushing over PRs.  No one should be doing anything critical with your PRs, so there is no harm in rewriting their history.  The "don't force push to publicly available branches!" thing that you are taught so often with Git is WAY over played.  It should really be "Don't force push to branches that you have reason to believe others have based work on".

2.  We don't mandate squashing.  In fact, in my mind, a few well ordered commits that show the progression of a new feature are desirable over a single pristine commit that comes out of no where.  We only ask people to squash if their commit history is really messy and it doesn't add anything of value.

That said: I think your commit history is fine... although @permcody is the ultimate decider on that front :smile: 

Thanks for all this work!  To my best knowledge I think this is the first time an external contributor has added a whole new "System" in MOOSE!  Definitely a landmark occasion!  This is a good one too... LOTS of people will use this for many years to come (including myself!).

Thanks!

lindsayad commented:
Thanks for the kind words :-) I needed this for my own work, and it was a fantastic learning experience (still ongoing of course!), so I'm glad I embarked on the journey.

Good to know about the force pushing of PRs. Yea, at this point I think I've already squashed most of the commits that would show a good "evolution" of the functionality of the PR. Any commits that are just "clean-up" I will go ahead and squash (already done for today's "clean-up" commits).

aeslaughter commented on framework/include/base/NonlinearSystem.h:
```diff
@@ -515,6 +517,7 @@ class NonlinearSystem : public SystemTempl<TransientNonlinearImplicitSystem>,
   KernelWarehouse _kernels;
   MooseObjectWarehouse<ScalarKernel> _scalar_kernels;
   MooseObjectWarehouse<DGKernel> _dg_kernels;
+  MooseObjectWarehouse<InterfaceKernel> _interface_kernels;
```
You made me very happy by using this without you needing to write a new warehouse class.

aeslaughter commented:
@lindsayad I would also be awesome if you add a wiki page [here](http://mooseframework.org/wiki/MooseSystems/) explaining the system, we will add it to our training slides after you add the content. Thanks.

aeslaughter commented:
@lindsayad I would also be awesome if you add a wiki page [here](http://mooseframework.org/wiki/MooseSystems/) explaining the system, we will add it to our training slides after you add the content. Thanks.

lindsayad commented:
I'll definitely do that if/when the PR gets merged.

lindsayad commented:
@friedmud Have you had a chance to look and see whether your thoughts have been addressed?

permcody commented:
ping @friedmud

permcody commented:
@lindsayad - Even if @friedmud signs off on this, it looks you'll need to rebase and resolve conflicts once more...

friedmud commented:
Yep: still planning on looking at this. Sorry for the delay.
On Mon, Jan 18, 2016 at 12:13 PM Cody Permann <notifications@github.com>
wrote:

> @lindsayad <https://github.com/lindsayad> - Even if @friedmud
> <https://github.com/friedmud> signs off on this, it looks you'll need to
> rebase and resolve conflicts once more...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-172594670>.
>

lindsayad commented:
Just wanted to note that the merge conflicts have been fixed (at the present time)

permcody commented:
Thanks @lindsayad. @friedmud - we probably don't want to wait 2 more weeks on this. Maybe you can squeeze a few mins in to look at this?

friedmud commented:
Yes.  I should have time tonight.

friedmud commented:
I haven't tested it yet... But I did look through it and the code changes look reasonable.

However: I think the new objects should have their own I put file syntax and their own sub directory.  This is really a whole new system and may be used by people who aren't doing anything associated with DG.

What do you guys think?

permcody commented:
While that may be true, Alex has been at this PR for over two months. It may be time to accept it and move forward with a new system if that's what everyone would like.

lindsayad commented:
Ok, I think I ran into an issue. So far, I have been using constant material properties for my InterfaceKernels and everything has worked fine. Now, however, I'm trying to access a neighbor material property on the slave side of the interface that is a function of variables on the slave side. However, when computeQpProperties gets called through reinitMaterialsNeighbor in my ComputeResidualThread::onInterface method, the coupled values of the variables in the Material Class don't match the initial conditions I set. In fact the variable values are zero.

In my material class I set my coupled variables like so:

  _em(isCoupled("em") ? coupledValue("em") : _zero)

The variable "em" exists and is non-zero on the slave side, but does not exist on the master side. So it almost seems like I'm trying to access the slave side material value using the master side value for "em". I haven't been able to track down how to fix this yet. Continuing to work on it, but if anyone can point me in a good direction, that would be great.

lindsayad commented:
Think I'm narrowing it down... Looks like I need to make the neighbor material object generated in FEProblem::addMaterial have have the boolean flag _coupleable_neighbor = true

lindsayad commented:
Alright cool, I got the neighbor material to get the right coupled variables (I think this could also have been a bug for DGKernels) by adding the middle line in FEProblem::addMaterial:

      current_parameters.set<Moose::MaterialDataType>("_material_data_type") = Moose::NEIGHBOR_MATERIAL_DATA;
      current_parameters.set<bool>("_neighbor") = true;
      object_name = name + "_neighbor";

@friedmud How much work is it to set up input file syntax?

permcody commented:
Alex,

It's not a whole lot of work but there are a few steps and some new
terminology. MOOSE uses an Action system to build new objects so you'd have
to make an Action (which is very trivial by the way) that creates the
object through the factory and adds it to the FEProblem. There are several
examples of these in the "actions" folder inside of the framework. Next you
have to create a task for the purpose of deciding when these Actions occur.
The tasks are just names that describe a step in the setup process, nothing
more. You create a new name and then place it in the master ordering list
which can be found in ../base/Moose.C. After that you create an association
from your task to your Action (just another method call) also in
../base/Moose.C. Finally, you have to pick the Syntax which is associated
with the Action which goes in ../src/parser/MooseSyntax.C.

None of those steps are difficult but there are four of them to weed
through. Yes we have an Example (#15) plus a few slides in the
MooseWorkshop material to help you out.

If you'd like to take a crack at it, that'd be great but I was hoping to
get this in there before going down that path. What do you think?

Cody

On Tue, Jan 26, 2016 at 8:23 AM Alex Lindsay <notifications@github.com>
wrote:

> Alright cool, I got the neighbor material to get the right coupled
> variables (I think this could also have been a bug for DGKernels) by adding
> the middle line in FEProblem::addMaterial:
>
>   current_parameters.set<Moose::MaterialDataType>("_material_data_type") = Moose::NEIGHBOR_MATERIAL_DATA;
>   current_parameters.set<bool>("_neighbor") = true;
>   object_name = name + "_neighbor";
>
> @friedmud <https://github.com/friedmud> How much work is it to set up
> input file syntax?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175073968>.
>

lindsayad commented:
I would be all for getting this merged ASAP :-)

If/when we do go down that path, would any difficulty arise from the 
fact that InterfaceKernel inherits from a base class (DGKernel)?

On 01/26/2016 10:47 AM, Cody Permann wrote:
> Alex,
>
> It's not a whole lot of work but there are a few steps and some new
> terminology. MOOSE uses an Action system to build new objects so you'd 
> have
> to make an Action (which is very trivial by the way) that creates the
> object through the factory and adds it to the FEProblem. There are several
> examples of these in the "actions" folder inside of the framework. 
> Next you
> have to create a task for the purpose of deciding when these Actions 
> occur.
> The tasks are just names that describe a step in the setup process, 
> nothing
> more. You create a new name and then place it in the master ordering list
> which can be found in ../base/Moose.C. After that you create an 
> association
> from your task to your Action (just another method call) also in
> ../base/Moose.C. Finally, you have to pick the Syntax which is associated
> with the Action which goes in ../src/parser/MooseSyntax.C.
>
> None of those steps are difficult but there are four of them to weed
> through. Yes we have an Example (#15) plus a few slides in the
> MooseWorkshop material to help you out.
>
> If you'd like to take a crack at it, that'd be great but I was hoping to
> get this in there before going down that path. What do you think?
>
> Cody
>
> On Tue, Jan 26, 2016 at 8:23 AM Alex Lindsay <notifications@github.com>
> wrote:
>
> > Alright cool, I got the neighbor material to get the right coupled
> > variables (I think this could also have been a bug for DGKernels) by 
> adding
> > the middle line in FEProblem::addMaterial:
> >
> > 
> current_parameters.set<Moose::MaterialDataType>("_material_data_type") 
> = Moose::NEIGHBOR_MATERIAL_DATA;
> > current_parameters.set<bool>("_neighbor") = true;
> > object_name = name + "_neighbor";
> >
> > @friedmud <https://github.com/friedmud> How much work is it to set up
> > input file syntax?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-175073968>.
> >
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175085219>.
>

permcody commented:
It actually has nothing to do with inheritance or anything else. The names
in the input file are just whatever we'd like them to be. That's by design
so end users can customize the syntax however they wish. It's fairly
straightforward regardless.

On Tue, Jan 26, 2016 at 9:09 AM Alex Lindsay <notifications@github.com>
wrote:

> I would be all for getting this merged ASAP :-)
>
> If/when we do go down that path, would any difficulty arise from the
> fact that InterfaceKernel inherits from a base class (DGKernel)?
>
> On 01/26/2016 10:47 AM, Cody Permann wrote:
> > Alex,
> >
> > It's not a whole lot of work but there are a few steps and some new
> > terminology. MOOSE uses an Action system to build new objects so you'd
> > have
> > to make an Action (which is very trivial by the way) that creates the
> > object through the factory and adds it to the FEProblem. There are
> several
> > examples of these in the "actions" folder inside of the framework.
> > Next you
> > have to create a task for the purpose of deciding when these Actions
> > occur.
> > The tasks are just names that describe a step in the setup process,
> > nothing
> > more. You create a new name and then place it in the master ordering list
> > which can be found in ../base/Moose.C. After that you create an
> > association
> > from your task to your Action (just another method call) also in
> > ../base/Moose.C. Finally, you have to pick the Syntax which is associated
> > with the Action which goes in ../src/parser/MooseSyntax.C.
> >
> > None of those steps are difficult but there are four of them to weed
> > through. Yes we have an Example (#15) plus a few slides in the
> > MooseWorkshop material to help you out.
> >
> > If you'd like to take a crack at it, that'd be great but I was hoping to
> > get this in there before going down that path. What do you think?
> >
> > Cody
> >
> > On Tue, Jan 26, 2016 at 8:23 AM Alex Lindsay <notifications@github.com>
> > wrote:
> >
> > > Alright cool, I got the neighbor material to get the right coupled
> > > variables (I think this could also have been a bug for DGKernels) by
> > adding
> > > the middle line in FEProblem::addMaterial:
> > >
> > >
> > current_parameters.set<Moose::MaterialDataType>("_material_data_type")
> > = Moose::NEIGHBOR_MATERIAL_DATA;
> > > current_parameters.set<bool>("_neighbor") = true;
> > > object_name = name + "_neighbor";
> > >
> > > @friedmud <https://github.com/friedmud> How much work is it to set up
> > > input file syntax?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/idaholab/moose/pull/5943#issuecomment-175073968>.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-175085219>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175094187>.
>

lindsayad commented:
That's nice. I will plan on implementing that at some point, but if you 
and the development team are willing to take the PR in its current 
state, that would be my preference.

On 01/26/2016 11:18 AM, Cody Permann wrote:
> It actually has nothing to do with inheritance or anything else. The names
> in the input file are just whatever we'd like them to be. That's by design
> so end users can customize the syntax however they wish. It's fairly
> straightforward regardless.
>
> On Tue, Jan 26, 2016 at 9:09 AM Alex Lindsay <notifications@github.com>
> wrote:
>
> > I would be all for getting this merged ASAP :-)
> >
> > If/when we do go down that path, would any difficulty arise from the
> > fact that InterfaceKernel inherits from a base class (DGKernel)?
> >
> > On 01/26/2016 10:47 AM, Cody Permann wrote:
> > > Alex,
> > >
> > > It's not a whole lot of work but there are a few steps and some new
> > > terminology. MOOSE uses an Action system to build new objects so you'd
> > > have
> > > to make an Action (which is very trivial by the way) that creates the
> > > object through the factory and adds it to the FEProblem. There are
> > several
> > > examples of these in the "actions" folder inside of the framework.
> > > Next you
> > > have to create a task for the purpose of deciding when these Actions
> > > occur.
> > > The tasks are just names that describe a step in the setup process,
> > > nothing
> > > more. You create a new name and then place it in the master 
> ordering list
> > > which can be found in ../base/Moose.C. After that you create an
> > > association
> > > from your task to your Action (just another method call) also in
> > > ../base/Moose.C. Finally, you have to pick the Syntax which is 
> associated
> > > with the Action which goes in ../src/parser/MooseSyntax.C.
> > >
> > > None of those steps are difficult but there are four of them to weed
> > > through. Yes we have an Example (#15) plus a few slides in the
> > > MooseWorkshop material to help you out.
> > >
> > > If you'd like to take a crack at it, that'd be great but I was 
> hoping to
> > > get this in there before going down that path. What do you think?
> > >
> > > Cody
> > >
> > > On Tue, Jan 26, 2016 at 8:23 AM Alex Lindsay 
> <notifications@github.com>
> > > wrote:
> > >
> > > > Alright cool, I got the neighbor material to get the right coupled
> > > > variables (I think this could also have been a bug for DGKernels) by
> > > adding
> > > > the middle line in FEProblem::addMaterial:
> > > >
> > > >
> > > current_parameters.set<Moose::MaterialDataType>("_material_data_type")
> > > = Moose::NEIGHBOR_MATERIAL_DATA;
> > > > current_parameters.set<bool>("_neighbor") = true;
> > > > object_name = name + "_neighbor";
> > > >
> > > > @friedmud <https://github.com/friedmud> How much work is it to 
> set up
> > > > input file syntax?
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > 
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175073968>.
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/idaholab/moose/pull/5943#issuecomment-175085219>.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-175094187>.
> >
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175098305>.
>

friedmud commented:
I would rather not merge stuff until it's right. We'll never get back
around to fixing it.... and people will immediately start using it with the
way it is.... meaning that it will be a hassle to switch later.

I agree that we shouldn't force Alex to do it though. He's already done
plenty! It's something I can do, but I don't know if I'll have time to do
it in the next couple of days.

Is there a reason to rush this?
On Tue, Jan 26, 2016 at 11:29 AM Alex Lindsay <notifications@github.com>
wrote:

> That's nice. I will plan on implementing that at some point, but if you
> and the development team are willing to take the PR in its current
> state, that would be my preference.
>
> On 01/26/2016 11:18 AM, Cody Permann wrote:
> > It actually has nothing to do with inheritance or anything else. The
> names
> > in the input file are just whatever we'd like them to be. That's by
> design
> > so end users can customize the syntax however they wish. It's fairly
> > straightforward regardless.
> >
> > On Tue, Jan 26, 2016 at 9:09 AM Alex Lindsay <notifications@github.com>
> > wrote:
> >
> > > I would be all for getting this merged ASAP :-)
> > >
> > > If/when we do go down that path, would any difficulty arise from the
> > > fact that InterfaceKernel inherits from a base class (DGKernel)?
> > >
> > > On 01/26/2016 10:47 AM, Cody Permann wrote:
> > > > Alex,
> > > >
> > > > It's not a whole lot of work but there are a few steps and some new
> > > > terminology. MOOSE uses an Action system to build new objects so
> you'd
> > > > have
> > > > to make an Action (which is very trivial by the way) that creates the
> > > > object through the factory and adds it to the FEProblem. There are
> > > several
> > > > examples of these in the "actions" folder inside of the framework.
> > > > Next you
> > > > have to create a task for the purpose of deciding when these Actions
> > > > occur.
> > > > The tasks are just names that describe a step in the setup process,
> > > > nothing
> > > > more. You create a new name and then place it in the master
> > ordering list
> > > > which can be found in ../base/Moose.C. After that you create an
> > > > association
> > > > from your task to your Action (just another method call) also in
> > > > ../base/Moose.C. Finally, you have to pick the Syntax which is
> > associated
> > > > with the Action which goes in ../src/parser/MooseSyntax.C.
> > > >
> > > > None of those steps are difficult but there are four of them to weed
> > > > through. Yes we have an Example (#15) plus a few slides in the
> > > > MooseWorkshop material to help you out.
> > > >
> > > > If you'd like to take a crack at it, that'd be great but I was
> > hoping to
> > > > get this in there before going down that path. What do you think?
> > > >
> > > > Cody
> > > >
> > > > On Tue, Jan 26, 2016 at 8:23 AM Alex Lindsay
> > <notifications@github.com>
> > > > wrote:
> > > >
> > > > > Alright cool, I got the neighbor material to get the right coupled
> > > > > variables (I think this could also have been a bug for DGKernels)
> by
> > > > adding
> > > > > the middle line in FEProblem::addMaterial:
> > > > >
> > > > >
> > > >
> current_parameters.set<Moose::MaterialDataType>("_material_data_type")
> > > > = Moose::NEIGHBOR_MATERIAL_DATA;
> > > > > current_parameters.set<bool>("_neighbor") = true;
> > > > > object_name = name + "_neighbor";
> > > > >
> > > > > @friedmud <https://github.com/friedmud> How much work is it to
> > set up
> > > > > input file syntax?
> > > > >
> > > > > —
> > > > > Reply to this email directly or view it on GitHub
> > > > >
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-175073968>.
> > > > >
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <https://github.com/idaholab/moose/pull/5943#issuecomment-175085219
> >.
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/idaholab/moose/pull/5943#issuecomment-175094187>.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5943#issuecomment-175098305>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-175102858>.
>

lindsayad commented:
No reason to rush on my behalf. The only thing I'm leery of is the PR growing stale and having conflicts with the base branch. I don't want to have to do any more substantial re-writes if it's avoidable

lindsayad commented:
Currently on vacation...so I've decided to try and tackle the action system. My first implementation resulted in this error:

*** ERROR ***
Task add_interface_kernel is not registered to build DGKernel derived objects

Recall that InterfaceKernel is a child of DGKernel. Now for this first attempted implementation, I didn't include a params.registerBase statement in the InterfaceKernel parameters construction (which is obviously wrong). When I add it in, it produces a segmentation fault. The fact that something bad happened in this latter case didn't surprise me since I have a params.registerBase statement in both DGKernel (the parent) and InterfaceKernel (child) parameter constructions. Any ideas? Don't see any ways to "changeBase".

Currently building a debug executable right now

permcody commented:
I don't know if we have another situation like that in the framework. Let
us know what you see in the debug exec.
On Mon, Feb 8, 2016 at 6:24 PM Alex Lindsay <notifications@github.com>
wrote:

> Currently on vacation...so I've decided to try and tackle the action
> system. My first implementation resulted in this error:
>
> *** ERROR ***
> Task add_interface_kernel is not registered to build DGKernel derived
> objects
>
> Recall that InterfaceKernel is a child of DGKernel. Now for this first
> attempted implementation, I didn't include a params.registerBase statement
> in the InterfaceKernel parameters construction (which is obviously wrong).
> When I add it in, it produces a segmentation fault. The fact that something
> bad happened in this latter case didn't surprise me since I have a
> params.registerBase statement in both DGKernel (the parent) and
> InterfaceKernel (child) parameter constructions. Any ideas? Don't see any
> ways to "changeBase".
>
> Currently building a debug executable right now
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-181658679>.
>

lindsayad commented:
Well the problem was actually in another piece of code I had added, but this new line in InterfaceKernel appears to make things work:

  params.set<std::string>("_moose_base") = "InterfaceKernel";

jwpeterson commented:
> Currently on vacation...so I've decided to try and tackle the action system. 

Worst. Vacation. Ever?

permcody commented:
Interesting... I'm not sure I love that workaround but you are in new
territory. We can add a better interface to support this new usage.

Cody

P.S. Definitely the worst vacation ever
On Mon, Feb 8, 2016 at 7:00 PM Alex Lindsay <notifications@github.com>
wrote:

> Well the problem was actually in another piece of code I had added, but
> this new line in InterfaceKernel appears to make things work:
>
> params.setstd::string("_moose_base") = "InterfaceKernel";
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5943#issuecomment-181667416>.
>

lindsayad commented:
Hahaha there's only so much fun I can have at a family reunion in 65+ Port Charlotte, Florida.

Well, InterfaceKernels now have their own directory and input file syntax. Hopefully, a merge isn't too far down the road...

friedmud commented:
Ok - this looks good to me.  Anything more can be done later.

This is great work @lindsayad !  Thanks so much for your time and effort here!

permcody commented:
Well @lindsayad - After 136 comments on this thread we are going to finally merge this...

YaqiWang commented:
I will be the first one saying: Congratulations! ;-)

lindsayad commented:
Awesome. Thanks for all the help and patience guys! Very proud and happy to contribute to such a great project :-)

You guys can be the judge (and this should have been in the first line of my commit message), but I was always hoping that the work here could close #6041 and #6042 as well

Alex
